File: 2022_solutions/typescript/src/Signals.ts
const count_unique_characters = (input_string: string): number => {
    const set = new Set<string>();

    for (let index = 0; index < input_string.length; index++) {
        set.add(input_string[index]);
    }

    return set.size;
}

const start_of_packet_marker = (input_string: string): number => {
    for (let index = 0; index < input_string.length - 3; index++) {
        const substring = input_string.substring(index, index + 4);
        const count = count_unique_characters(substring);

        if (count == 4) {
            return index + 4;
        }
    }

    return 0;
}

const start_of_message_marker = (input_string: string): number => {
    for (let index = 0; index < input_string.length - 13; index++) {
        const substring = input_string.substring(index, index + 14);
        const count = count_unique_characters(substring);

        if (count == 14) {
            return index + 14;
        }
    }

    return 0;
}

export {start_of_packet_marker, start_of_message_marker};

File: 2022_solutions/rust/src/signals.rs
use std::collections::HashSet;

pub fn start_of_packet_marker(input_string: &str) -> usize {
    for index in 0..input_string.len() - 3 {
        let count = input_string.chars()
            .skip(index)
            .take(4)
            .collect::<HashSet<char>>()
            .len();

        if count == 4 {
            return index + 4;
        }
    }

    return 0;
}

pub fn start_of_message_marker(input_string: &str) -> usize {
    for index in 0..input_string.len() - 13 {
        let count = input_string.chars()
            .skip(index)
            .take(14)
            .collect::<HashSet<char>>()
            .len();

        if count == 14 {
            return index + 14;
        }
    }

    return 0;
}